panic on NewStoreFromGraphdriver

Reported by @justincormack 

```
time="2016-01-19T14:51:15.264108533Z" level=info msg="Layer sha256:4e0e182967f847e292d3b8c8ebff2cf567c6a38713ddbad9f40e8dad2e603ff4 cleaned up" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x815856]

goroutine 1 [running]:
github.com/docker/docker/layer.NewStoreFromGraphDriver(0x7fd78c5eee38, 0xc820616d80, 0x7fd78c5eed38, 0xc820357220, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/layer/layer_store.go:90 +0x446
github.com/docker/docker/layer.NewStoreFromOptions(0xc820611800, 0xf, 0xc820258140, 0x20, 0x7ffd15737edb, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/layer/layer_store.go:66 +0x496
github.com/docker/docker/daemon.NewDaemon(0xc82032f080, 0xc8203580e0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/daemon.go:707 +0xe36
main.(*DaemonCli).CmdDaemon(0xc820329540, 0xc82000a1a0, 0x4, 0x4, 0x0, 0x0)
	/go/src/github.com/docker/docker/docker/daemon.go:252 +0x16e6
reflect.callMethod(0xc82027d530, 0xc8206b1c48)
	/usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a1a0, 0x4, 0x4, 0x1, 0xc82027d530, 0x0, 0x0, 0xc82027d530, 0x0, 0x474114, ...)
	/usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc82027d4a0, 0xc82000a190, 0x5, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
	/go/src/github.com/docker/docker/docker/docker.go:63 +0x43c

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 51 [chan receive]:
github.com/docker/docker/daemon.setupDumpStackTrap.func1(0xc82013eb40)
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:17 +0x4d
created by github.com/docker/docker/daemon.setupDumpStackTrap
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:20 +0x12a

goroutine 18 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 50 [select, locked to thread]:
runtime.gopark(0x19ea8c8, 0xc820021f28, 0x16ebe38, 0x6, 0x41c018, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc820021f28, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc820021f28)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x815856]

goroutine 1 [running]:
github.com/docker/docker/layer.NewStoreFromGraphDriver(0x7f8e3873eb38, 0xc8205a7bf0, 0x7f8e3873ea38, 0xc820079b80, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/layer/layer_store.go:90 +0x446
github.com/docker/docker/layer.NewStoreFromOptions(0xc8206a3840, 0xf, 0xc8202958e0, 0x20, 0x7ffe86ce0ed6, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/layer/layer_store.go:66 +0x496
github.com/docker/docker/daemon.NewDaemon(0xc82007a840, 0xc8203180c0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/daemon.go:707 +0xe36
main.(*DaemonCli).CmdDaemon(0xc820294f00, 0xc82000a1a0, 0x4, 0x4, 0x0, 0x0)
	/go/src/github.com/docker/docker/docker/daemon.go:252 +0x16e6
reflect.callMethod(0xc8206accf0, 0xc8206bfc48)
	/usr/local/go/src/reflect/value.go:628 +0x1fc
reflect.methodValueCall(0xc82000a1a0, 0x4, 0x4, 0x1, 0xc8206accf0, 0x0, 0x0, 0xc8206accf0, 0x0, 0x474114, ...)
	/usr/local/go/src/reflect/asm_amd64.s:29 +0x36
github.com/docker/docker/cli.(*Cli).Run(0xc8206acc60, 0xc82000a190, 0x5, 0x5, 0x0, 0x0)
	/go/src/github.com/docker/docker/cli/cli.go:89 +0x383
main.main()
	/go/src/github.com/docker/docker/docker/docker.go:63 +0x43c

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 6 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 36 [select, locked to thread]:
runtime.gopark(0x19ea8c8, 0xc8205a5728, 0x16ebe38, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc8205a5728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc8205a5728)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 37 [chan receive]:
github.com/docker/docker/daemon.setupDumpStackTrap.func1(0xc82036a600)
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:17 +0x4d
created by github.com/docker/docker/daemon.setupDumpStackTrap
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:20 +0x12a
```

--- a/layer/filestore.go
+++ b/layer/filestore.go
@@ -10,6 +10,7 @@ import (
 	"path/filepath"
 	"regexp"
 	"strconv"
+	"strings"
 
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/distribution/digest"
@@ -154,7 +155,7 @@ func (fms *fileMetadataStore) GetParent(layer ChainID) (ChainID, error) {
 		return "", err
 	}
 
-	dgst, err := digest.ParseDigest(string(content))
+	dgst, err := digest.ParseDigest(strings.TrimSpace(string(content)))
 	if err != nil {
 		return "", err
 	}
@@ -168,7 +169,7 @@ func (fms *fileMetadataStore) GetDiffID(layer ChainID) (DiffID, error) {
 		return "", err
 	}
 
-	dgst, err := digest.ParseDigest(string(content))
+	dgst, err := digest.ParseDigest(strings.TrimSpace(string(content)))
 	if err != nil {
 		return "", err
 	}
@@ -177,16 +178,17 @@ func (fms *fileMetadataStore) GetDiffID(layer ChainID) (DiffID, error) {
 }
 
 func (fms *fileMetadataStore) GetCacheID(layer ChainID) (string, error) {
-	content, err := ioutil.ReadFile(fms.getLayerFilename(layer, "cache-id"))
+	contentBytes, err := ioutil.ReadFile(fms.getLayerFilename(layer, "cache-id"))
 	if err != nil {
 		return "", err
 	}
+	content := strings.TrimSpace(string(contentBytes))
 
-	if !stringIDRegexp.MatchString(string(content)) {
+	if !stringIDRegexp.MatchString(content) {
 		return "", errors.New("invalid cache id value")
 	}
 
-	return string(content), nil
+	return content, nil
 }
 
 func (fms *fileMetadataStore) TarSplitReader(layer ChainID) (io.ReadCloser, error) {
@@ -227,32 +229,34 @@ func (fms *fileMetadataStore) SetMountParent(mount string, parent ChainID) error
 }
 
 func (fms *fileMetadataStore) GetMountID(mount string) (string, error) {
-	content, err := ioutil.ReadFile(fms.getMountFilename(mount, "mount-id"))
+	contentBytes, err := ioutil.ReadFile(fms.getMountFilename(mount, "mount-id"))
 	if err != nil {
 		return "", err
 	}
+	content := strings.TrimSpace(string(contentBytes))
 
-	if !stringIDRegexp.MatchString(string(content)) {
+	if !stringIDRegexp.MatchString(content) {
 		return "", errors.New("invalid mount id value")
 	}
 
-	return string(content), nil
+	return content, nil
 }
 
 func (fms *fileMetadataStore) GetInitID(mount string) (string, error) {
-	content, err := ioutil.ReadFile(fms.getMountFilename(mount, "init-id"))
+	contentBytes, err := ioutil.ReadFile(fms.getMountFilename(mount, "init-id"))
 	if err != nil {
 		if os.IsNotExist(err) {
 			return "", nil
 		}
 		return "", err
 	}
+	content := strings.TrimSpace(string(contentBytes))
 
-	if !stringIDRegexp.MatchString(string(content)) {
+	if !stringIDRegexp.MatchString(content) {
 		return "", errors.New("invalid init id value")
 	}
 
-	return string(content), nil
+	return content, nil
 }
 
 func (fms *fileMetadataStore) GetMountParent(mount string) (ChainID, error) {
@@ -264,7 +268,7 @@ func (fms *fileMetadataStore) GetMountParent(mount string) (ChainID, error) {
 		return "", err
 	}
 
-	dgst, err := digest.ParseDigest(string(content))
+	dgst, err := digest.ParseDigest(strings.TrimSpace(string(content)))
 	if err != nil {
 		return "", err
 	}

--- a/layer/layer_store.go
+++ b/layer/layer_store.go
@@ -86,6 +86,7 @@ func NewStoreFromGraphDriver(store MetadataStore, driver graphdriver.Driver) (St
 		l, err := ls.loadLayer(id)
 		if err != nil {
 			logrus.Debugf("Failed to load layer %s: %s", id, err)
+			continue
 		}
 		if l.parent != nil {
 			l.parent.referenceCount++
@@ -109,22 +110,22 @@ func (ls *layerStore) loadLayer(layer ChainID) (*roLayer, error) {
 
 	diff, err := ls.store.GetDiffID(layer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get diff id for %s: %s", layer, err)
 	}
 
 	size, err := ls.store.GetSize(layer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get size for %s: %s", layer, err)
 	}
 
 	cacheID, err := ls.store.GetCacheID(layer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get cache id for %s: %s", layer, err)
 	}
 
 	parent, err := ls.store.GetParent(layer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get parent for %s: %s", layer, err)
 	}
 
 	cl = &roLayer{

